impl/src/main/java/com/okta/sdk/impl/client/DefaultClientBuilder.java

```diff
@@ -307,18 +325,63 @@ public Client build() {
             this.cacheManager = cacheManagerBuilder.build();
         }
 
-        if (this.clientConfig.getClientCredentialsResolver() == null && this.clientCredentials != null) {
-            this.clientConfig.setClientCredentialsResolver(new DefaultClientCredentialsResolver(this.clientCredentials));
+        if (isOAuth2Flow()) {
+            this.clientConfig.setClientCredentialsResolver(new OAuth2ClientCredentialsResolver(this.clientCredentials));
         }
-        else if (this.clientConfig.getClientCredentialsResolver() == null) {
-            this.clientConfig.setClientCredentialsResolver(new DefaultClientCredentialsResolver(clientConfig));
+        else {
+            if (this.clientConfig.getClientCredentialsResolver() == null && this.clientCredentials != null) {
+                this.clientConfig.setClientCredentialsResolver(new DefaultClientCredentialsResolver(this.clientCredentials));
+            } else if (this.clientConfig.getClientCredentialsResolver() == null) {
+                this.clientConfig.setClientCredentialsResolver(new DefaultClientCredentialsResolver(this.clientConfig));
+            }
         }
 
         if (this.clientConfig.getBaseUrlResolver() == null) {
             ConfigurationValidator.assertOrgUrl(this.clientConfig.getBaseUrl(), allowNonHttpsForTesting);
             this.clientConfig.setBaseUrlResolver(new DefaultBaseUrlResolver(this.clientConfig.getBaseUrl()));
         }
 
+        // OAuth2
+        if (isOAuth2Flow()) {
+            this.clientConfig.setAuthenticationScheme(AuthenticationScheme.OAUTH2);
+            accessTokenRetrieverService = new AccessTokenRetrieverServiceImpl(clientConfig);
+
+            // Get access token asynchronously
+            CompletableFuture<OAuth2AccessToken> completableFuture = CompletableFuture.supplyAsync(() -> {
+                OAuth2AccessToken accessToken;
+                try {
+                    accessToken = accessTokenRetrieverService.getOAuth2AccessToken();
+                } catch (IOException | InvalidKeyException e) {
+                    log.error("Exception occurred:", e);
+                    throw new IllegalArgumentException("Exception occurred", e);
+                } catch (HttpException e) {
+                    log.error("Exception occurred:", e);
+                    throw e;
+                }
+
+                return accessToken;
+            }).whenComplete((accessTokenResult, e) -> {
+                if (e != null) {
+                    log.error("Exception occurred:", e);
+                    throw new RuntimeException(e);
+                }
+            });
+
+            try {
+                OAuth2AccessToken oAuth2AccessToken = completableFuture.get();
```

Comment:
This blocks, so threading out above probably doesn't help much, and would add some potential problems.
(NOTE: this might not matter once this logic is removed from this class)